src/main/java/org/opensearch/ad/ml/ModelManager.java

```diff
@@ -944,102 +944,122 @@ public void getTotalUpdates(String modelId, String detectorId, ActionListener<Lo
 
     /**
      * Compute anomaly result for the given data point
-     * @param detectorId Detector Id
      * @param datapoint Data point
-     * @param entityName entity's name like "server_1"
      * @param modelState the state associated with the entity
      * @param modelId the model Id
+     * @param detector Detector accessor
+     * @param entity entity accessor
+     *
      * @return anomaly result, confidence, and the corresponding RCF score.
      */
     public ThresholdingResult getAnomalyResultForEntity(
-        String detectorId,
         double[] datapoint,
-        String entityName,
         ModelState<EntityModel> modelState,
-        String modelId
+        String modelId,
+        AnomalyDetector detector,
+        Entity entity
     ) {
-        ThresholdingResult result = null;
-
         if (modelState != null) {
-            EntityModel model = modelState.getModel();
-            Queue<double[]> samples = model.getSamples();
-            samples.add(datapoint);
-            if (samples.size() > this.rcfNumMinSamples) {
-                samples.remove();
+            EntityModel entityModel = modelState.getModel();
+
+            if (entityModel == null) {
+                entityModel = new EntityModel(entity, new ArrayDeque<>(), null, null);
+                modelState.setModel(entityModel);
             }
 
-            result = maybeTrainBeforeScore(modelState, entityName);
+            // trainModelFromExistingSamples may be able to make models not null
+            if (entityModel.getRcf() == null || entityModel.getThreshold() == null) {
+                entityColdStarter.trainModelFromExistingSamples(modelState);
+            }
+
+            if (entityModel.getRcf() != null && entityModel.getThreshold() != null) {
+                return score(datapoint, modelId, modelState);
+            } else {
```

Comment:
Else is unnecessary because if always returns. Any reason to keep it?

Reply:
Mostly for readability.  Some discussions here: https://stackoverflow.com/questions/3261849/why-is-else-rarely-used-after-if-x-then-return